src/main/java/build/buf/protovalidate/internal/evaluator/ConstraintResolver.java

```diff
@@ -35,9 +36,13 @@ class ConstraintResolver {
    * @param desc the message descriptor.
    * @return the resolved {@link MessageConstraints}.
    */
-  MessageConstraints resolveMessageConstraints(Descriptor desc)
+  MessageConstraints resolveMessageConstraints(Descriptor desc, ExtensionRegistry registry)
       throws InvalidProtocolBufferException, CompilationException {
     DescriptorProtos.MessageOptions options = desc.getOptions();
+    // If the protovalidate message extension is unknown, reparse using extension registry.
+    if (options.getUnknownFields().hasField(ValidateProto.message.getNumber())) {
+      options = DescriptorProtos.MessageOptions.parseFrom(options.toByteString(), registry);
```

Comment:
This is the key fix for handling dynamic messages created from a file descriptor set. If we determine that the protovalidate extension is in the unknown fields, we'll reparse it using our extension registry.